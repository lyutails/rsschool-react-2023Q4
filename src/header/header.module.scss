.header {
  height: 100px;
  width: 100%;
  background: linear-gradient(0.75turn, white, black);

  &_wrapper {
    width: 90vw;
    margin: 0 auto;
    display: flex;
    flex-direction: row;
    align-items: center;
  }

  &_logic {
    width: 80vw;
    height: 80px;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
  }

  &_input_wrapper {
    display: flex;
    flex-direction: column;
    width: 30vw;
    margin-right: 20px;
  }

  &_input {
    height: 30px;

    &:focus {
      outline: none;
    }
  }

  &_under_input {
    color: white;
  }
}

.header_input:focus::placeholder {
  color: transparent;
}

.grogu_looking {
  width: 100px;
  height: 100px;
  position: absolute;
  right: 0;
  top: 0;
  background-image: url('../assets/grogu_looking.png');
  background-repeat: no-repeat;
  background-size: contain;
  background-blend-mode: lighten;
}
