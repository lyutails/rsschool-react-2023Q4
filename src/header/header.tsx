import React, { useState } from 'react';
import style from '../app.module.scss';
import { useNavigate } from 'react-router-dom';

interface Props {
  searchValue: string;
  fetchSpecies: (inputValue: string) => void;
  changeSearchValue: (data: string) => void;
  changePage: (data: number) => void;
}

export function Header({
  searchValue,
  fetchSpecies,
  changeSearchValue,
  changePage,
}: Props) {
  const [hasError, setHasError] = useState<boolean>(false);
  const navigate = useNavigate();

  if (hasError) {
    throw new Error('lalala');
  }
  return (
    <div className={style.header}>
      <div className={style.header_wrapper}>
        <div className={style.header_logic}>
          <input
            placeholder="Search..."
            value={searchValue}
            className={style.header_input}
            onChange={(e): void => {
              changeSearchValue(e.target.value);
              e.target.value.length
                ? localStorage.setItem('searchValue', `${e.target.value}`)
                : localStorage.removeItem('searchValue');
              changePage(1);
              const query = e.target.value.length
                ? `/page/1/search/${e.target.value}`
                : `/page/1`;
              navigate(query);
            }}
          ></input>
          <button
            className={style.cross}
            onClick={() => {
              localStorage.removeItem('searchValue');
              changeSearchValue('');
              fetchSpecies('');
              changePage(1);
              navigate(`/page/1`);
            }}
          >
            cross
          </button>
          <button
            className={style.header_search}
            onClick={() => {
              fetchSpecies(searchValue);
            }}
          ></button>
          <button
            className={style.header_error}
            onClick={() => setHasError(true)}
          >
            Get Error
          </button>
        </div>
      </div>
      <div className={style.grogu_looking}></div>
    </div>
  );
}
