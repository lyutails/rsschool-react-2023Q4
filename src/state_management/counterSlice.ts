import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

export interface CounterState {
  value: number;
}

const initialState: CounterState = {
  value: 1,
};

export const counterSlice = createSlice({
  name: 'counter',
  initialState,
  reducers: {
    totalSpeciesFound: (state, action: PayloadAction<number>) => {
      state.value = action.payload;
    },
  },
});

export const { totalSpeciesFound } = counterSlice.actions;

export default counterSlice.reducer;
