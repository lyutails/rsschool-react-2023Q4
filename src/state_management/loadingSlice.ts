import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

const initialState = {
  value: false,
};

export const loadingSlice = createSlice({
  name: 'loading',
  initialState,
  reducers: {
    getLoading: (state, action: PayloadAction<boolean>) => {
      state.value = action.payload;
    },
  },
});

export const { getLoading } = loadingSlice.actions;

export default loadingSlice.reducer;
