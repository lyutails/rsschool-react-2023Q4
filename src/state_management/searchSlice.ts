import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

export interface SearchState {
  value: string;
}

const initialState: SearchState = {
  value: '',
};

export const searchSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    makeSearch: (state, action: PayloadAction<string>) => {
      state.value = action.payload;
    },
    clearSearch: (state) => {
      state.value = '';
    },
  },
});

export const { makeSearch, clearSearch } = searchSlice.actions;

export default searchSlice.reducer;
