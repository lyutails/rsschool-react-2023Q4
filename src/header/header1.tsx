/* import { beforeEach, describe, expect, test, vi, it } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { Header } from './header';

describe('testing header component', () => {
  const inputTestId = 'lalala';
  const inputValue = 'yo';
  const onChangeInput = vi.fn();
  let input: HTMLInputElement;
  beforeEach(() => {
    render(
      <Header
        data-testid={inputTestId}
        value={inputValue}
        onChange={onChangeInput}
      ></Header>
    );
    input = screen.getByTestId(inputTestId);
  });

  test('check default input value', () => {
    expect(input).toBe(inputValue);
  });

  it('should empty local storage', () => {
    render(<Header />);
    const crossButton = screen.getByText('cross');
    fireEvent.click(crossButton);
    expect(localStorage).toContain('');
  });
});
 */
