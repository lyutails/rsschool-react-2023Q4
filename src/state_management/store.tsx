import { configureStore } from '@reduxjs/toolkit';
import counterReducer from './counterSlice';
import speciesReducer from './speciesSlice';
import searchSlice from './searchSlice';
import { speciesApi } from '../api/api';
import { setupListeners } from '@reduxjs/toolkit/query';

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    species: speciesReducer,
    search: searchSlice,
    [speciesApi.reducerPath]: speciesApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(speciesApi.middleware),
});

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
