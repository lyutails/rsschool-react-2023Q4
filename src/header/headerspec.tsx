/* import { describe, it, expect, vi } from 'vitest';
import { Header } from './header';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { store } from '../state_management/store';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

describe('check what is going on in header', () => {
  it('header has a button', () => {
    const changePage = vi.fn((data: number) => data);
    render(
      <Provider store={store}>
        <Header changePage={changePage} />
      </Provider>,
      { wrapper: BrowserRouter }
    );
    const button = screen.getByRole('button', { name: 'Get Error' });
    expect(button).toBeInTheDocument();
  });
  it('a button text content is cross', () => {
    const changePage = function (data: number) {
      return data;
    };
    render(
      <Provider store={store}>
        <Header changePage={changePage} />
      </Provider>,
      { wrapper: BrowserRouter }
    );
    const button = screen.getByText(/Cross/);
    expect(button.textContent).toBe('Cross');
  });
  it('a button text content is Search and Save', () => {
    const changePage = function (data: number) {
      return data;
    };
    render(
      <Provider store={store}>
        <Header changePage={changePage} />
      </Provider>,
      { wrapper: BrowserRouter }
    );
    const button = screen.getByText(/Search and Save/);
    expect(button.textContent).toBe('Search and Save');
  });
  it('a button text content is Search and Save', () => {
    const changePage = function (data: number) {
      return data;
    };
    render(
      <Provider store={store}>
        <Header changePage={changePage} />
      </Provider>,
      { wrapper: BrowserRouter }
    );
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'lalala' } });
    userEvent.type(input, 'lalala');
  });
});
 */