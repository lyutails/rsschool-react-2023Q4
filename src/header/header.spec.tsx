import { describe, it, expect, vi } from 'vitest';
import { Header } from './header';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { store } from '../state_management/store';
import { BrowserRouter } from 'react-router-dom';

describe('check what is going on in header', () => {
  it('header has a button', () => {
    const changePage = vi.fn((data: number) => data);
    render(
      <Provider store={store}>
        <Header changePage={changePage} />
      </Provider>,
      { wrapper: BrowserRouter }
    );
    const button = screen.getByRole('button', { name: 'Get Error' });
    expect(button).toBeInTheDocument();
  });
  it('a button text content is cross', () => {
    const changePage = function (data: number) {
      return data;
    };
    render(
      <Provider store={store}>
        <Header changePage={changePage} />
      </Provider>,
      { wrapper: BrowserRouter }
    );
    const button = screen.getByText(/Cross/);
    expect(button.textContent).toBe('Cross');
  });
});
