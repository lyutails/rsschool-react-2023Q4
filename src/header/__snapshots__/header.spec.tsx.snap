// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`check what is going on in header > The component UI should not change 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_header_bf677c"
      >
        <div
          class="_header_wrapper_bf677c"
        >
          <div
            class="_header_logic_bf677c"
          >
            <div
              class="_header_input_wrapper_bf677c"
            >
              <input
                class="_header_input_bf677c"
                placeholder="Search..."
                value=""
              />
              <div
                class="_header_under_input_bf677c"
              >
                Save to store: 
              </div>
            </div>
            <button
              class="_cross_bf677c"
            >
              Cross
            </button>
            <button
              class="_header_search_bf677c"
              value=""
            >
              Search and Save
            </button>
            <button
              class="_header_error_bf677c"
            >
              Get Error
            </button>
          </div>
        </div>
        <div
          class="_grogu_looking_bf677c"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_header_bf677c"
    >
      <div
        class="_header_wrapper_bf677c"
      >
        <div
          class="_header_logic_bf677c"
        >
          <div
            class="_header_input_wrapper_bf677c"
          >
            <input
              class="_header_input_bf677c"
              placeholder="Search..."
              value=""
            />
            <div
              class="_header_under_input_bf677c"
            >
              Save to store: 
            </div>
          </div>
          <button
            class="_cross_bf677c"
          >
            Cross
          </button>
          <button
            class="_header_search_bf677c"
            value=""
          >
            Search and Save
          </button>
          <button
            class="_header_error_bf677c"
          >
            Get Error
          </button>
        </div>
      </div>
      <div
        class="_grogu_looking_bf677c"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
